core/src/main/java/org/bitcoinj/protocols/payments/PaymentSession.java

```diff
@@ -304,35 +305,39 @@ public SendRequest getSendRequest() {
      * NOTE: This does not broadcast the transactions to the bitcoin network, it merely sends a Payment message to the
      * merchant confirming the payment.
      * Returns an object wrapping PaymentACK once received.
-     * If the PaymentRequest did not specify a payment_url, returns null and does nothing.
+     * If the PaymentRequest did not specify a payment_url, completes exceptionally.
```

Comment:
payment_url is optional, so I think completing exceptionally is not the right thing to do.

Reply:
But then again in this state of the protocol, maybe it is an exception.

Reply:
`payment_url` may be optional, but this method call is `sendPayment` and it is supposed to generate a payment request, send it to the server, and return an ACK. So sending trying to send a payment when there is no URL to send it to seems exceptional to me.

And, in the general case of a `CompletableFuture` it is supposed to either:

a. Complete with a value
b. Complete with an exception

A method that returns a `CompletableFuture` should generally not return `null` because they are supposed to be chain able, etc. I made some changes elsewhere and they were broken by this method returning `null`. I have never seen a method that is supposed to return a `CompletableFuture` (or `Future`) return `null`.

So if this method can't return a Future that completes with an ACK it should complete with an exception.

Reply:
(This may be preaching to the choir after your second message, but I thought you might appreciate the full rationale)

Reply:
To drone on a little bit further, the current `sendPayment` has 4 possible behaviors:

1. Returns a Completable future that complete with an ACK
2. Returns a Completable future that completes exceptionally
3. Returns `null`
4. Throws an exception

After this PR only (1) or (2) should occur.

Reply:
@schildbach is this explanation good enough to consider this issue resolved?

Reply:
> this method call is `sendPayment` and it is supposed to generate a payment request,

It's sending a **payment** message, rather than a **payment request** message, right?

Reply:
> @schildbach is this explanation good enough to consider this issue resolved?

I guess yes (-:

Reply:
> > this method call is `sendPayment` and it is supposed to generate a payment request,
> 
> It's sending a **payment** message, rather than a **payment request** message, right?

Yes. I edited the above to say: ~~request~~ message

I confusing used the word "request" as network-level terminology, as in request/response protocol.